Adding support for MUMPS sparse linear solver

zanellia opened the pull request:
I am working on a package that relies on the Schur complement method in qpOASES and having to rely on MA57 makes it more cumbersome for people to install it. For this reason I have added support for the freely redistributable solver MUMPS. 

In order to use MUMPS it define `USE_SOLVER = MUMPS`, e.g., in make_linux.mk. The Makefile will obtain the source code of MUMPS and compile and install it to `<qpOASESroot>external/mumps_installation/lib` automatically. In order to run qpOASES, `<qpOASESroot>external/mumps_installation/lib` needs to be present in `LD_LIBRARY_PATH`.

The tests are still not passing, however the solver takes exactly the same iterations until MUMPS fails to compute a factorization (at least for test_qrecipeSchur). 

Also, I am not sure who does code reviews at the moment. Maybe @jferreau or @apotschka?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/coin-or/qpOASES?pullRequest=133) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/coin-or/qpOASES?pullRequest=133) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/coin-or/qpOASES?pullRequest=133) it.</sub>

zanellia commented:
I am not sure who does code reviews at the moment, Maybe @jferreau or @apotschka?

zanellia commented:
This is the output of `./qrecipeSchur` using MA57 and MUMPS, respectively. I am printing the iterates and they are equivalent until iteration 91 and then the residuals of the iterative refinement takes very large values (~1e12). That happens around line 31501 (the files can be easily compared with, e.g., vimdiff).

[ma57.log](https://github.com/coin-or/qpOASES/files/9418001/ma57.log)
[mumps.log](https://github.com/coin-or/qpOASES/files/9418002/mumps.log)

zanellia commented:
In order to investigate the issue, I have added a script to generate the data for sparse QPs from Python. I have added to the repo the data for 2 QPs and example `qrecipeSchur.cpp` can be run with with different data commenting/uncommenting https://github.com/zanellia/qpOASES/blob/mumps/examples/qrecipeSchur.cpp#L40. Surprisingly, for the a trivial QP, the dense and sparse versions of qpOASES return different results:

`Primal error (dense and sparse):  2.45e+02`
while (with MUMPS)
`Primal error (dense and Schur):   2.71e-08`

with MA57, I obtain a slightly better accuracy for the Schur complement method:
`Primal error (dense and sparse):  2.45e+02` (obviously)
`Primal error (dense and Schur):   4.55e-14`

In conclusion, I would be tempted to say that the interface to MUMPS seems to work OK - MUMPS is just not very accurate in some cases. I was going to update the tests, but I am not sure how to proceed given the fact that the nullspace method returns a very different solution.

apotschka commented on examples/qrecipeSchur.cpp:
```diff
@@ -110,18 +110,19 @@ int main( )
     Bounds guessedBounds( NV );
     guessedBounds.init(NV);
     for (int i = 0; i < NV; i++) {
-        guessedBounds.setupBound(i, (SubjectToStatus)0);
```
Why activate all constraints? This usually leads to rank deficiency of working set constraint Jacobian.

zanellia replied:
I guess I was trying to obtain a non empty Schur complement starting from iteration 0 for debugging purposes. Anyway, this part of the example has been reverted.

apotschka commented on examples/qrecipeSchur.cpp:
```diff
@@ -110,18 +110,19 @@ int main( )
 
     printf("fine till here\n");
-	qrecipeSchur.init(H, g, A, lb, ub, lbA, ubA, nWSR, 0, NULL, NULL, &guessedBounds, &guessedConstraints);
+	// qrecipeSchur.init(H, g, A, lb, ub, lbA, ubA, nWSR, 0, NULL, NULL, &guessedBounds, &guessedConstraints);
```
Why not use guessed constraints?

apotschka commented on make_linux.mk:
```diff
@@ -77,7 +77,7 @@ else ifeq ($(USE_SOLVER), MA27)
 	LINKHSL =
 else ifeq ($(USE_SOLVER), MUMPS)
-	LIB_SOLVER = /home/andrea/feasible_sqp/external/mumps_installation/lib/libcoinmumps.so
+	LIB_SOLVER = $(PWD)/../external/mumps_installation/lib/libcoinmumps.so
```
OK!

apotschka commented on make_linux.mk:
```diff
@@ -39,12 +39,6 @@ BINDIR = ${TOP}/bin
 
 # Matlab include directory (ADAPT TO YOUR LOCAL SETTINGS!)
 #MATLAB_IDIR   = ${HOME}/Programs/matlab/extern/include/
-# MATLAB_IDIR   = /usr/local/matlab/extern/include/
```
OK!

apotschka commented on testing/cpp/test_qrecipeSchur.cpp:
```diff
@@ -98,10 +98,34 @@ int main( )
 
 	/* solve with sparse matrices (Schur complement) */
 	#ifndef SOLVER_NONE
+	Options options;
```
Fix indentation.

zanellia replied:
reverted.

apotschka reviewed (commented):
Huge thanks, Andrea. Your contribution is greatly appreciated. This really fills a wide-open gap!

I have made a couple comments that I would like to ask you to address before merging the pull request. Some of them on earlier commits you have already addressed yourself in later commits.

Most importantly:
 - We need to be able to change the C and Fortran integer data types via `__USE_LONG_INTEGERS__` and `__USE_LONG_FINTS__`.
 - Does it still compile on Windows and MacOS? If you cannot answer that yet, we might still complete the merge and open two tickets to check that.

apotschka commented:
> In order to investigate the issue, I have added a script to generate the data for sparse QPs from Python. I have added to the repo the data for 2 QPs and example `qrecipeSchur.cpp` can be run with with different data commenting/uncommenting https://github.com/zanellia/qpOASES/blob/mumps/examples/qrecipeSchur.cpp#L40. Surprisingly, for the a trivial QP, the dense and sparse versions of qpOASES return different results:
> 
> `Primal error (dense and sparse): 2.45e+02` while (with MUMPS) `Primal error (dense and Schur): 2.71e-08`
> 
> with MA57, I obtain a slightly better accuracy for the Schur complement method: `Primal error (dense and sparse): 2.45e+02` (obviously) `Primal error (dense and Schur): 4.55e-14`
> 
> In conclusion, I would be tempted to say that the interface to MUMPS seems to work OK - MUMPS is just not very accurate in some cases. I was going to update the tests, but I am not sure how to proceed given the fact that the nullspace method returns a very different solution.

I did not have time yet to investigate this issue further.

zanellia commented:
> Huge thanks, Andrea. Your contribution is greatly appreciated. This really fills a wide-open gap!
> 
> I have made a couple comments that I would like to ask you to address before merging the pull request. Some of them on earlier commits you have already addressed yourself in later commits.
> 


Thanks for the detailed code review @apotschka! I addressing your comments today. On a high level,

> Most importantly:
> 
>     * We need to be able to change the C and Fortran integer data types via `__USE_LONG_INTEGERS__` and `__USE_LONG_FINTS__`. 

I had troubles getting qpOASES running with some version of MA57 (especially running tests of my solver on Travis CI, the MATLAB version used on the server would affect the MA57 version which in turn caused erratic behavior with, e.g., segfaults). So the change is actually related to MA57 rather than MUMPS. I will try to get to a more precise conclusion regarding what breaks, but for now I can definitely remove it from this PR.

> 
>     * Does it still compile on Windows and MacOS? If you cannot answer that yet, we might still complete the merge and open two tickets to check that.

Unfortunately, I do not have access to a macOS machine at the moment. Could this be an additional reason to setup , e.g., Travis CI for the repo? If that's useful, I can look into it.

apotschka commented:
> Unfortunately, I do not have access to a macOS machine at the moment. Could this be an additional reason to setup , e.g., Travis CI for the repo? If that's useful, I can look into it.

The addition of Travis CI would be really nice for future developments. If you have time to set it up, please go for it. If you need extended access rights here on GitHub, please let me know.

zanellia commented on make_linux.mk:
```diff
@@ -68,6 +69,10 @@ else ifeq ($(USE_SOLVER), MA27)
 	DEF_SOLVER = SOLVER_MA27
 	LINKHSL =
+else ifeq ($(USE_SOLVER), MUMPS)
+	LIB_SOLVER = /home/andrea/feasible_sqp/external/mumps_installation/lib/libcoinmumps.so
```
Already resolved within PR.

zanellia commented on src/SparseSolver.cpp:
```diff
@@ -45,15 +45,17 @@ void MyPrintf(const char* pformat, ... );
 #if SOLVER_MUMPS
 
-#include "mumps_compat.h"
+#define USE_COMM_WORLD -987654
```
This seems to be the standard way of using the default MPI Fortran communicator (http://mumps.enseeiht.fr/doc/userguide_5.5.1.pdf, Section 9.2). It's also used in Ipopt's source code https://github.com/coin-or/Ipopt/blob/stable/3.14/src/Algorithm/LinearSolvers/IpMumpsSolverInterface.cpp#L55.

zanellia commented on src/SparseSolver.cpp:
```diff
@@ -45,15 +45,17 @@ void MyPrintf(const char* pformat, ... );
-#ifdef COIN_USE_MUMPS_MPI_H
-#include "mumps_mpi.h"
-#else
+#include <../../mumps_installation/include/coin-or/mumps/mumps_compat.h>
```
Addressed here [06bdf06](https://github.com/coin-or/qpOASES/pull/133/commits/06bdf060a4bbb801fdafa320eff52f59d6afea72)

zanellia commented on make_linux.mk:
```diff
@@ -39,18 +39,24 @@ BINDIR = ${TOP}/bin
 
 # system or replacement BLAS/LAPACK
-REPLACE_LINALG = 1
+REPLACE_LINALG = 0
```
accidental, resolved here [44a07a9](https://github.com/coin-or/qpOASES/pull/133/commits/44a07a94f23ed594e260fe34361ab8f2c28425e9)

zanellia commented on make_linux.mk:
```diff
@@ -115,7 +122,7 @@ CPPFLAGS = -Wall -pedantic -Wshadow -Wfloat-equal -O3 -Wconversion -Wsign-conver
 
 # libraries to link against when building qpOASES .so files
-LINK_LIBRARIES = ${LIB_LAPACK} ${LIB_BLAS} -lm ${LIB_SOLVER}
+LINK_LIBRARIES = ${LIB_LAPACK} ${LIB_BLAS} -lm ${LIB_SOLVER} -ldl
```
Unfortunately, I have not tested this on macOS/Windows. We could set up Travis CI to facilitate cross-platform testing.

zanellia replied:
Unfortunately, I haven't tested this yet.

zanellia commented on make_linux.mk:
```diff
@@ -39,18 +39,24 @@ BINDIR = ${TOP}/bin
 #MATLAB_IDIR   = ${HOME}/Programs/matlab/extern/include/
-MATLAB_IDIR   = /usr/local/matlab/extern/include/
-MATLAB_LIBDIR = /usr/local/matlab/bin/glnxa64/
+# MATLAB_IDIR   = /usr/local/matlab/extern/include/
```
These changes are not in final PR.

zanellia commented on make_linux.mk:
```diff
@@ -59,10 +65,11 @@ endif
 ifeq ($(USE_SOLVER), MA57)
-	LIB_SOLVER = ${MATLAB_LIBDIR}/libmwma57.so
+	# LIB_SOLVER = ${MATLAB_LIBDIR}/libmwma57.so
+    LIB_SOLVER = /home/andrea/feasible_sqp/external/coin_hsl_source/.libs/libcoinhsl.so
```
already resolved by later commits.

zanellia commented on src/SparseSolver.cpp:
```diff
@@ -1122,73 +1122,6 @@ static void MPIfini(void)
 }
 #endif /* !USE_MPI_H */
 
-extern "C"
```
This does not affect MA57, since it's in the constructor for MumpsSparseSolver, right?

zanellia commented on examples/qrecipeSchur.cpp:
```diff
@@ -93,7 +93,35 @@ int main( )
+	Options options;
+	// options.setToDefault();
+	// options.printLevel = PL_TABULAR;
+	options.printLevel = PL_HIGH;
```
reverted.

zanellia commented on src/SparseSolver.cpp:
```diff
@@ -1069,6 +1085,540 @@ returnValue Ma57SparseSolver::copy( 	const Ma57SparseSolver& rhs
+/*****************************************************************************
+ *****************************************************************************
+ *****************************************************************************
+ *  M U M P S P A R E S E S O L V E R                                        *
```
addressed in [0988b25](https://github.com/coin-or/qpOASES/pull/133/commits/0988b2591dfa7db2829d7f689b24548bc728abc5)

zanellia commented on include/qpOASES/Types.hpp:
```diff
@@ -174,8 +174,10 @@ typedef double real_t;
-typedef unsigned long uint_t;
+// typedef long int_t;
+// typedef unsigned long uint_t;
+typedef int int_t;
```
I have reverted the commit - it was part of an (sloppy) attempt to make qpOASES work with different versions of MA57. With some versions I would get segfaults with it.

zanellia commented on include/qpOASES/Types.hpp:
```diff
@@ -184,7 +186,8 @@ typedef unsigned int uint_t;
 
 /** Defines FORTRAN integer type. Might be platform dependent! */
 #ifdef __USE_LONG_FINTS__
-typedef long fint_t;
```
As above, I was trying to make qpOASES work with different versions of MA57. Surely not the right way to address it, but  why would you pass `long` integers (64 bytes on an x86) to MA57 API which takes Fortran integers (32 bits)?

zanellia commented on include/qpOASES/SparseSolver.hpp:
```diff
@@ -434,23 +436,92 @@ class MumpsSparseSolver: public SparseSolver
         void* mumps_ptr_;           /** Primary MUMPS data structure */
 
-		fint_t dim;				    /**< Dimension of the current linear system. */
+		int dim;				    /**< Dimension of the current linear system. */
```
reverted.

apotschka commented:
> 2. I reverted the changes to the definitions of `int_t` and `fint_t`. I am still not sure why you'd need to pass `long` integers (64 bits) to MA57's API which takes Fortran integers (32 bit). This was actually causing segfaults with some versions of MA57. That's why I added that hack (and then forgot to clean it up).

Matlab ships with an MA57 with long integers. If you want to run qpOASES from within Matlab with MA57, you need to take care of that.

> 3. The qrecipe_Schur example and associated test run fine after removing `options.enableEqualities = BT_TRUE`. Not sure why. If you or someone else has a possible explanation I would be interested :)
> 4. Unfortunately the smallSchur test is still failing.

Unfortunately, I do not have time right now to look into these issues. Number 3 sounds like problems with almost singular matrices and different behavior of different linear solvers in that case.

I'll merge the pull request now and hope potential bugs will be fixed later in a community effort.
